allennlp/tests/nn/beam_search_test.py

```diff
@@ -0,0 +1,171 @@
+              state: Dict[str, torch.Tensor]) -> Tuple[torch.Tensor, Dict[str, torch.Tensor]]:
+    """
+    Take decoding step.
+
+    This is a simple function that defines how probabilities are computed for the
+    next time step during the beam search.
+
+    We use a simple target vocabulary of size 6. In this vocabulary, index 0 represents
+    the start token, and index 5 represents the end token. The transition probability
+    from a state where the last predicted token was token `j` to new token `i` is
+    given by the `(i, j)` element of the matrix `transition_probabilities`.
+    """
+    log_probs_list = []
+    for last_token in last_predictions:
+        log_probs = torch.log(transition_probabilities[last_token.item()])
+        log_probs_list.append(log_probs)
+
+    return torch.stack(log_probs_list), state
+
+
+class BeamSearchTest(AllenNlpTestCase):
+
+    def setUp(self):
+        super(BeamSearchTest, self).setUp()
+        self.end_index = transition_probabilities.size()[0] - 1
+        self.beam_search = BeamSearch(self.end_index, max_steps=10, beam_size=3)
+
+        # This is what the top k should look like for each item in the batch.
+        self.expected_top_k = np.array(
+                [[1, 2, 3, 4, 5],
+                 [2, 3, 4, 5, 5],
+                 [3, 4, 5, 5, 5]]
+        )
+
+        # This is what the log probs should look like for each item in the batch.
+        self.expected_log_probs = np.log(np.array([0.4, 0.3, 0.2]))
+
+    def _check_results(self,
+                       batch_size: int = 5,
+                       expected_top_k: np.array = None,
+                       expected_log_probs: np.array = None,
+                       beam_search: BeamSearch = None,
+                       state: Dict[str, torch.Tensor] = None,
+                       expected_finished_state: Dict[str, np.array] = None) -> None:
+        expected_top_k = expected_top_k if expected_top_k is not None else self.expected_top_k
+        expected_log_probs = expected_log_probs if expected_log_probs is not None else self.expected_log_probs
+        state = state or {}
+        expected_finished_state = expected_finished_state or {}
+
+        beam_search = beam_search or self.beam_search
+        beam_size = beam_search.beam_size
+
+        initial_predictions = torch.tensor([0] * batch_size)  # pylint: disable=not-callable
+        top_k, log_probs = beam_search.search(initial_predictions, state, take_step)  # type: ignore
+
+        # top_k should be shape `(batch_size, beam_size, max_predicted_length)`.
+        assert list(top_k.size())[:-1] == [batch_size, beam_size]
+        np.testing.assert_array_equal(top_k[0].numpy(), expected_top_k)
+
+        # log_probs should be shape `(batch_size, beam_size, max_predicted_length)`.
+        assert list(log_probs.size()) == [batch_size, beam_size]
+        np.testing.assert_allclose(log_probs[0].numpy(), expected_log_probs)
+
+        # check finished state.
```

Comment:
If you're only testing this once, I'd pull the logic out into the test itself. I try to keep test helpers dead simple to avoid surprises and improve readability.